### PR TITLE
Rewrite deployer in etherlime-lib in TypeScript

### PR DESCRIPTION
**Pull request by desimira**

### CODE REVIEW

File: packages/etherlime-lib/src/deployed-contract/contract-at.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
-const ethers = require('ethers');
-const { ganacheSetupConfig } = require('etherlime-config');
-const { isSigner, isProvider } = require('etherlime-utils');
+import { ethers, Wallet } from 'ethers';
```
[the remark this answers is not in the document]
Comment (Perseverance) [reply packages/etherlime-lib/src/deployed-contract/contract-at.ts]: Also, due to the import of wallet, we should substitute `ethers.Wallet` with just `Wallet`

File: packages/etherlime-lib/src/deployed-contract/etherlime-ganache-wrapper.ts [hunk 2]
```diff
@@ -15,54 +18,58 @@ class EtherlimeGanacheWrapper extends DeployedContractWrapper {
...
+			const accSigner = new Wallet(acc.secretKey, provider);
+			const accContract = new Contract(contractAddress, contract.abi, accSigner);
 			this.instances.push(accContract)
 			this.instancesMap[accSigner.address] = accContract
 		}
 	}
 
-	from(addressOrSignerOrIndex) {
-		if (Number.isInteger(addressOrSignerOrIndex)) {
+	from(addressOrSignerOrIndex: string | Wallet & etherlimeWallet | number): Contract {
```
Comment (Perseverance) [inline packages/etherlime-lib/src/deployed-contract/etherlime-ganache-wrapper.ts]: I do not think it should be `&`

File: packages/etherlime-lib/src/deployer/deployer.ts [hunk 3]
```diff
@@ -13,8 +19,13 @@ class Deployer {
 	 * @param {*} provider ethers.provider instance
 	 * @param {*} defaultOverrides [Optional] default deployment overrides
 	 */
-	constructor(signer, provider, defaultOverrides) {
-		this._validateInput(signer, defaultOverrides);
+
+	signer: Wallet;
+	provider: JsonRpcProvider | InfuraProvider | Web3Provider;
```
Comment (Perseverance) [inline packages/etherlime-lib/src/deployer/deployer.ts]: I think, InfuraProvider is a subclass of JSONRPC Provider so you do not need to have Infura here too

Comment (Perseverance) [reply packages/etherlime-lib/src/deployer/deployer.ts]: Not sure but the same may apply for Web3Provider too

File: packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts [hunk 4]
```diff
@@ -13,46 +15,47 @@ class EtherlimeGanacheDeployer extends JSONRPCDeployer {
...
 
-	static _validatePortInput(port) {
+	static _validatePortInput(port: number): void {
 		if (!isNumber(port)) {
 			throw new Error(`Passed port (${port}) is not valid port`);
 		}
 	}
 
-	toString() {
+	toString(): string {
```
Comment (Perseverance) [inline packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: For some class methods you've added access modifiers (public, private, etc). I think it is good idea to either keep putting them in all classes or none (I prefer the first)

File: packages/etherlime-lib/src/types/types.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    secretKey: string,
+	signer: Wallet
+}
+
+export interface JSONRPCGlobal extends Global {
+  coverageSubprovider: any
+  provider: any
+}
+
+export interface Generic<T> {
```
Comment (Perseverance) [inline packages/etherlime-lib/src/types/types.ts]: General question - If we rewrite the etherlime-utils in typescript, where will this interface end up, here or in the utils?

File: packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts (hunk 4 above)
Comment (ochikov) [reply packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: And just keep in mind that if we don't have a modifier, the default state is `public`.

Comment (desimira) [reply packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: Yep, I've put modifiers only where the default should be changed. Shouldn't this method "toString()" be public?

File: packages/etherlime-lib/src/types/types.ts (hunk 5 above)
Comment (desimira) [reply packages/etherlime-lib/src/types/types.ts]: This interface is universal and can be used for any type that needed to be customised. So it's not specially connected with the deployer, thus I think we can move it in 'etherlime-utils'.

File: packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts (hunk 4 above)
Comment (Perseverance) [reply packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: Most languages that have modifiers (like solidity) it is considered good practice to put explicitly the modifiers, just to form a habit of thinking of the access modification. It seems reasonable to me. What you guys think?

Comment (desimira) [reply packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: The reason I omit them is the principe of "if anything can be skipped - skip it" and also I have read somewhere that not adding "public" distinguishes  "protected" and "private" and it's easy to see quickly what's what. But I could also agree with you!

Comment (Perseverance) [reply packages/etherlime-lib/src/deployer/etherlime-ganache-deployer/etherlime-ganache-deployer.ts]: I agree with your point, lets skip them.

**Pull request merged**